src/libs/zmbv/zmbv.cpp

```diff
@@ -184,11 +184,11 @@ int VideoCodec::CompareBlock(const int vx, const int vy, const FrameBlock & bloc
 	int ret = 0;
 	P *pold = reinterpret_cast<P *>(oldframe) + block.start + (vy * pitch) + vx;
 	P *pnew = reinterpret_cast<P *>(newframe) + block.start;
-	;
+
 	for (auto y = 0; y < block.dy; y++) {
 		for (auto x = 0; x < block.dx; x++) {
-			const auto test = 0 - ((pold[x] - pnew[x]) & 0x00ffffff);
-			ret -= check_cast<int>(test >> 31);
+			const int test = 0 - ((pold[x] - pnew[x]) & 0x00ffffff);
+			ret -= (test >> 31);
 		}
```

Comment:
That's pretty subtle! Great find, @weirddan455.

Maybe the `ret = ` could be renamed to describe the number of pixels found to be different (if that's what we're counting?)

Also, can the "zero minus" and the `ret -=` be flipped over to positives, so it reads more naturally:

```c++
    int diff_count = 0;
    ...    
    for ( ...
        for ( ...
            diff_count += ((pold[x] ^ pnew[x]) & 0x00ffffff) != 0;
        }
    }

    return diff_count;
```

(if not, maybe a comment is needed to describe why we need to subtract negatives)

Reply:
Oh, yeah I guess it is just masking out the alpha channel (which is un-unsed) and then counting the number of pixels that differ.

I was wondering what that `test >> 31` meant but it's just checking the sign bit and incrementing the return value in the most obtuse way possible :sweat_smile: 

I'll stick your code in there and run a test to see if the values are identical.  If they are, yours is a lot more readable so I'll go with that.

Reply:
> it's just checking the sign bit and incrementing the return value in the most obtuse way possible

:100: I guess some of these elite coding patterns were for performance reasons (or something!) :hurtrealbad:

Reply:
Ok, yours wins.  Added a log statement that fired if the values differed from the existing code and it was the same in both EGA (16bpp) and VGA (32bpp).

The only thing I can figure is the original author was trying to avoid a branch that the `!= 0` might cause but I took a quick look at the assembly out of curiosity and couldn't tell much difference.  The compiler was pretty smart and inlined and auto-vectorized a lot of this code with SIMD.

Changed to match your suggestion for readability :smile:

Reply:
> The compiler was pretty smart and inlined and auto-vectorized a lot of this code with SIMD.

Nice!

Reply:
> > it's just checking the sign bit and incrementing the return value in the most obtuse way possible
> 
> 💯 . I guess some of these elite coding patterns were for performance reasons (or something!) :hurtrealbad:

Standard 90's game /demo coder fare -- nothing unusual to see, moving on 😎 

Plus yeah, optimising C compilers were in their infancy 20-30 years ago compared to what we have now. One way to make C code from the 80/90s fly with almost zero effort is to simply recompile it with clang/gcc instead or Watcom/Turbo C/etc. (if you are able to find your way around the DOS era nemory model bullshit in the modern compiler).